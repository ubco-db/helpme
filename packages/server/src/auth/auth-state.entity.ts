import {
  BaseEntity,
  Column,
  CreateDateColumn,
  Entity,
  JoinColumn,
  ManyToOne,
  PrimaryColumn,
} from 'typeorm';
import { OrganizationModel } from '../organization/organization.entity';

@Entity('auth_state_model')
export class AuthStateModel extends BaseEntity {
  @PrimaryColumn({ type: 'text' })
  state: string;

  @CreateDateColumn({ type: 'timestamp with time zone' })
  createdAt: Date;

  @Column({ type: 'integer', default: 60 })
  expiresIn: number;

  @Column({ type: 'integer', nullable: false })
  organizationId: number;

  @ManyToOne(() => OrganizationModel, (org) => org.userAuthStates, {
    onDelete: 'CASCADE',
  })
  @JoinColumn({ name: 'organizationId' })
  organization: OrganizationModel;
}
