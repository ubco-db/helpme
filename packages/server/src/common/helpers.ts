import { Request } from 'express';

export function getCookie(req: Request, cookieName: string): string {
  const cookieHeader = req.headers.cookie;
  if (!cookieHeader) {
    return null;
  }

  const cookies: string[] = cookieHeader.split(';');
  for (const cookie of cookies) {
    const [name, value] = cookie.split('=');
    if (name.trim() === cookieName) {
      return value;
    }
  }
}
