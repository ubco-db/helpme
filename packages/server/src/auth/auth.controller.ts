import {
  Controller,
  Get,
  HttpStatus,
  Param,
  Query,
  Req,
  Res,
  Post,
  Body,
  UseGuards,
} from '@nestjs/common';
import { Request, Response } from 'express';
import { ConfigService } from '@nestjs/config';
import { AuthService } from './auth.service';
import {
  AccountRegistrationParams,
  AccountType,
  ERROR_MESSAGES,
  PasswordRequestResetBody,
  PasswordRequestResetWithTokenBody,
  RegistrationTokenDetails,
} from '@koh/common';
import { JwtService } from '@nestjs/jwt';
import { OrganizationModel } from 'organization/organization.entity';
import { UserModel } from 'profile/user.entity';
import * as request from 'superagent';
import { MailService } from 'mail/mail.service';
import {
  TokenAction,
  TokenType,
  UserTokenModel,
} from 'profile/user-token.entity';
import { JwtAuthGuard } from 'guards/jwt-auth.guard';
import * as bcrypt from 'bcrypt';
import { getCookie } from '../common/helpers';

interface RequestUser {
  userId: string;
}

@Controller('auth')
export class AuthController {
  private GOOGLE_AUTH_URL = 'https://accounts.google.com/o/oauth2/v2/auth';

  constructor(
    private jwtService: JwtService,
    private configService: ConfigService,
    private mailService: MailService,
    private authService: AuthService,
  ) {}

  @Get('shibboleth/:oid')
  async shibbolethAuth(
    @Req() req: Request,
    @Res() res: Response,
    @Param('oid') organizationId: number,
  ): Promise<any> {
    const organization = await OrganizationModel.findOne({
      where: { id: organizationId },
    });

    if (!organization) {
      return res.redirect(`/auth/failed/40000`);
    }
    if (!organization.ssoEnabled) {
      return res.redirect(`/auth/failed/40002`);
    }

    const uid = req.headers['x-trust-auth-uid'] ?? null;
    const mail = req.headers['x-trust-auth-mail'] ?? null;
    const role = req.headers['x-trust-auth-role'] ?? null;
    const givenName = req.headers['x-trust-auth-givenname'] ?? null;
    const lastName = req.headers['x-trust-auth-lastname'] ?? null;

    if (!uid || !mail || !role || !givenName || !lastName) {
      return res.redirect(`/auth/failed/40001`);
    }

    try {
      const userId = await this.authService.loginWithShibboleth(
        String(mail),
        String(role),
        String(givenName),
        String(lastName),
        organizationId,
      );

      this.enter(req, res, userId);
    } catch (err) {
      return res.redirect(`/auth/failed/40001`);
    }
  }

  @Get('link/:method/:oid')
  auth(
    @Res() res: Response,
    @Param('method') auth_method: string,
    @Param('oid') organizationId: number,
  ): Response<{ redirectUri: string }> {
    res.cookie('organization.id', organizationId, {
      httpOnly: true,
      secure: this.isSecure(),
    });

    switch (auth_method) {
      case 'google':
        return res.status(200).send({
          redirectUri:
            `${this.GOOGLE_AUTH_URL}?client_id=${process.env.GOOGLE_CLIENT_ID}.apps.googleusercontent.com` +
            `&redirect_uri=${process.env.GOOGLE_REDIRECT_URI}&response_type=code&scope=openid%20profile%20email`,
        });
      default:
        return res
          .status(HttpStatus.BAD_REQUEST)
          .send({ message: 'Invalid auth method' });
    }
  }

  @Post('registration/verify')
  @UseGuards(JwtAuthGuard)
  async validateRegistrationToken(
    @Res() res: Response,
    @Req() req: Request,
    @Body() registrationTokenDetails: RegistrationTokenDetails,
  ): Promise<Response<void>> {
    const { token } = registrationTokenDetails;

    const emailToken = await UserTokenModel.findOne({
      where: {
        token,
        token_type: TokenType.EMAIL_VERIFICATION,
        token_action: TokenAction.ACTION_PENDING,
        user: { id: Number((req.user as RequestUser).userId) },
      },
      relations: ['user'],
    });

    if (!emailToken) {
      return res.status(HttpStatus.BAD_REQUEST).send({
        message: 'Verification code was not found or it is not valid',
      });
    }

    if (emailToken.expires_at < parseInt(new Date().getTime().toString())) {
      return res.status(HttpStatus.BAD_REQUEST).send({
        message: 'Verification code has expired',
      });
    }

    emailToken.token_action = TokenAction.ACTION_COMPLETE;
    emailToken.user.emailVerified = true;
    await emailToken.user.save();
    await emailToken.save();
    const cookie = getCookie(req, '__SECURE_REDIRECT');

    if (cookie) {
      const decodedCookie = decodeURIComponent(cookie);
      return res.status(HttpStatus.TEMPORARY_REDIRECT).send({
        redirectUri: `/course/${decodedCookie.split(',')[0]}/invite?code=${decodedCookie.split(',')[1]}`,
      });
    } else {
      return res.status(HttpStatus.ACCEPTED).send({
        message: 'Email verified',
      });
    }
  }

  @Get('/password/reset/validate/:token')
  async validatePasswordResetToken(
    @Res() res: Response,
    @Param('token') token: string,
  ): Promise<Response<void>> {
    const passwordToken = await UserTokenModel.findOne({
      where: {
        token,
        token_type: TokenType.PASSWORD_RESET,
        token_action: TokenAction.ACTION_PENDING,
      },
    });

    if (!passwordToken) {
      return res.status(HttpStatus.BAD_REQUEST).send({
        message: 'Password reset token was not found or it is not valid',
      });
    }

    if (passwordToken.expires_at < parseInt(new Date().getTime().toString())) {
      return res.status(HttpStatus.BAD_REQUEST).send({
        message: 'Password reset token has expired',
      });
    }

    return res.status(HttpStatus.ACCEPTED).send({
      message: 'Password reset token is valid',
    });
  }

  @Post('/password/reset/:token')
  async resetPasswordToken(
    @Body() body: PasswordRequestResetWithTokenBody,
    @Res() res: Response,
    @Param('token') token: string,
  ): Promise<Response<void>> {
    const { password, confirmPassword } = body;

    if (password !== confirmPassword) {
      return res.status(HttpStatus.BAD_REQUEST).send({
        message: 'Passwords do not match',
      });
    }

    const passwordToken = await UserTokenModel.findOne({
      where: {
        token,
        token_type: TokenType.PASSWORD_RESET,
        token_action: TokenAction.ACTION_PENDING,
      },
      relations: ['user'],
    });

    if (!passwordToken) {
      return res.status(HttpStatus.BAD_REQUEST).send({
        message: 'Password reset token was not found or it is not valid',
      });
    }

    if (passwordToken.expires_at < parseInt(new Date().getTime().toString())) {
      return res.status(HttpStatus.BAD_REQUEST).send({
        message: 'Password reset token has expired',
      });
    }

    passwordToken.token_action = TokenAction.ACTION_COMPLETE;
    passwordToken.expires_at = parseInt(new Date().getTime().toString());
    await passwordToken.save();

    const salt = await bcrypt.genSalt(10);
    const hashedPassword = await bcrypt.hash(password, salt);

    passwordToken.user.password = hashedPassword;
    await passwordToken.user.save();

    return res.status(HttpStatus.ACCEPTED).send({
      message: 'Password reset successful',
    });
  }

  @Post('/password/reset')
  async resetPassword(
    @Body() body: PasswordRequestResetBody,
    @Res() res: Response,
  ): Promise<Response<void>> {
    const { email, recaptchaToken, organizationId } = body;

    if (!recaptchaToken) {
      return res
        .status(HttpStatus.BAD_REQUEST)
        .send({ message: 'Invalid recaptcha token' });
    }

    const response = await request.post(
      `https://www.google.com/recaptcha/api/siteverify?secret=${process.env.PRIVATE_RECAPTCHA_SITE_KEY}&response=${recaptchaToken}`,
    );

    if (!response.body.success) {
      return res.status(HttpStatus.BAD_REQUEST).send({
        message: 'Recaptcha token invalid',
      });
    }

    const user = await UserModel.findOne({
      where: {
        email,
        organizationUser: { organizationId },
        accountType: AccountType.LEGACY,
      },
      relations: ['organizationUser'],
    });

    if (!user) {
      return res
        .status(HttpStatus.BAD_REQUEST)
        .send({ message: 'User not found' });
    }

    if (!user.emailVerified) {
      return res
        .status(HttpStatus.BAD_REQUEST)
        .send({ message: 'Email not verified' });
    }

    const resetLink = await this.authService.createPasswordResetToken(user);

    this.mailService.sendPasswordResetEmail(
      user.email,
      `${process.env.DOMAIN}/account/password/${resetLink}`,
    );

    return res.status(HttpStatus.ACCEPTED).send({
      message: 'Password reset email sent',
    });
  }

  @Post('register')
  async register(
    @Body() body: AccountRegistrationParams,
    @Res() res: Response,
  ): Promise<Response<void>> {
    const {
      firstName,
      lastName,
      email,
      password,
      confirmPassword,
      sid,
      organizationId,
      recaptchaToken,
    } = body;

    if (!recaptchaToken) {
      return res
        .status(HttpStatus.BAD_REQUEST)
        .send({ message: 'Invalid recaptcha token' });
    }

    const response = await request.post(
      `https://www.google.com/recaptcha/api/siteverify?secret=${process.env.PRIVATE_RECAPTCHA_SITE_KEY}&response=${recaptchaToken}`,
    );

    if (!response.body.success) {
      return res.status(HttpStatus.BAD_REQUEST).send({
        message: 'Recaptcha token invalid',
      });
    }

    if (firstName.trim().length < 1 || lastName.trim().length < 1) {
      return res
        .status(HttpStatus.BAD_REQUEST)
        .send({ message: 'First and last name must be at least 1 character' });
    }

    if (firstName.trim().length > 32 || lastName.trim().length > 32) {
      return res
        .status(HttpStatus.BAD_REQUEST)
        .send({ message: 'First and last name must be at most 32 characters' });
    }

    if (email.trim().length < 4 || email.trim().length > 64) {
      return res
        .status(HttpStatus.BAD_REQUEST)
        .send({ message: 'Email must be between 4 and 64 characters' });
    }

    if (password.trim().length < 6 || password.trim().length > 32) {
      return res
        .status(HttpStatus.BAD_REQUEST)
        .send({ message: 'Password must be between 6 and 32 characters' });
    }

    if (password !== confirmPassword) {
      return res
        .status(HttpStatus.BAD_REQUEST)
        .send({ message: 'Passwords do not match' });
    }

    const organization = await OrganizationModel.findOne({
      where: { id: organizationId },
    });

    if (!organization) {
      return res
        .status(HttpStatus.BAD_REQUEST)
        .send({ message: 'Organization not found' });
    }

    const user = await UserModel.findOne({ where: { email } });

    if (user) {
      return res
        .status(HttpStatus.BAD_REQUEST)
        .send({ message: 'Email already exists' });
    }

    if (sid) {
      const result = await this.authService.studentIdExists(
        sid,
        organizationId,
      );
      if (result) {
        return res
          .status(HttpStatus.BAD_REQUEST)
          .send({ message: 'Student ID already exists' });
      }
    }

    try {
      const userId = await this.authService.register(
        firstName,
        lastName,
        email,
        password,
        sid ? sid : -1,
        organizationId,
      );

      const authToken = await this.createAuthToken(userId);

      if (authToken === null || authToken === undefined) {
        return res.status(HttpStatus.INTERNAL_SERVER_ERROR).send({
          message: ERROR_MESSAGES.loginController.invalidTempJWTToken,
        });
      }

      return res
        .cookie('auth_token', authToken, {
          httpOnly: true,
          secure: this.isSecure(),
        })
        .send({ message: 'Account created' });
    } catch (err) {
      return res.status(HttpStatus.BAD_REQUEST).send({ message: err.message });
    }
  }

  @Get('callback/:method')
  async callback(
    @Res() res: Response,
    @Param('method') auth_method: string,
    @Query('code') auth_code: string,
    @Req() req: Request,
  ): Promise<Response<void>> {
    const organizationId = getCookie(req, 'organization.id');

    if (!organizationId) {
      res.redirect(`/auth/failed/40000`);
    } else {
      try {
        let payload: number;

        switch (auth_method) {
          case 'google':
            payload = await this.authService.loginWithGoogle(
              auth_code,
              Number(organizationId),
            );
            break;
          default:
            return res
              .status(HttpStatus.BAD_REQUEST)
              .send({ message: 'Invalid auth method' });
        }

        res.clearCookie('organization.id', {
          httpOnly: true,
          secure: this.isSecure(),
        });

        this.enter(req, res, payload);
      } catch (err) {
        res.redirect(`/auth/failed/40001`);
      }
    }
  }

  private async enter(req: Request, res: Response, userId: number) {
    const authToken = await this.createAuthToken(userId);

    if (authToken === null || authToken === undefined) {
      return res
        .status(HttpStatus.INTERNAL_SERVER_ERROR)
        .send({ message: ERROR_MESSAGES.loginController.invalidTempJWTToken });
    }

    const cookie = getCookie(req, '__SECURE_REDIRECT');
    let redirectUrl: string;

    if (cookie) {
      const decodedCookie = decodeURIComponent(cookie);
      redirectUrl = `/course/${decodedCookie.split(',')[0]}/invite?code=${decodedCookie.split(',')[1]}`;
    } else {
      redirectUrl = '/courses';
    }

    res
      .cookie('auth_token', authToken, {
        httpOnly: true,
        secure: this.isSecure(),
      })
      .redirect(HttpStatus.FOUND, redirectUrl);
  }

  private async createAuthToken(userId: number): Promise<string> {
    // Expires in 30 days
    return await this.jwtService.signAsync({
      userId,
      expiresIn: 60 * 60 * 24 * 30,
    });
  }

  private isSecure(): boolean {
    return this.configService.get<string>('DOMAIN').startsWith('https://');
  }
}
