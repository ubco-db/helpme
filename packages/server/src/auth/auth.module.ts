import { Module } from '@nestjs/common';
import { AuthController } from './auth.controller';
import { AuthService } from './auth.service';
import { JwtStrategy } from 'login/jwt.strategy';
import { JwtModule } from '@nestjs/jwt';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { MailService } from 'mail/mail.service';
import { MailModule } from 'mail/mail.module';
import { CourseService } from 'course/course.service';
import { RedisProfileService } from 'redisProfile/redis-profile.service';
import { ChatbotApiService } from 'chatbot/chatbot-api.service';
@Module({
  imports: [
    JwtModule.registerAsync({
      imports: [ConfigModule],
      inject: [ConfigService],
      useFactory: async (configService: ConfigService) => ({
        secret: configService.get('JWT_SECRET'),
      }),
    }),
    MailModule,
  ],
  controllers: [AuthController],
  providers: [
    JwtStrategy,
    AuthService,
    CourseService,
    RedisProfileService,
    ChatbotApiService,
  ],
  exports: [AuthService],
})
export class AuthModule {}
