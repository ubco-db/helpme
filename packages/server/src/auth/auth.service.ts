import { AccountType, OrganizationRole } from '@koh/common';
import {
  BadRequestException,
  Injectable,
  InternalServerErrorException,
} from '@nestjs/common';
import { OAuth2Client } from 'google-auth-library';
import { OrganizationUserModel } from 'organization/organization-user.entity';
import { UserModel } from 'profile/user.entity';
import * as bcrypt from 'bcrypt';
import { TokenType, UserTokenModel } from 'profile/user-token.entity';
import { MailService } from 'mail/mail.service';
import { ChatTokenModel } from 'chatbot/chat-token.entity';
import { v4 } from 'uuid';

@Injectable()
export class AuthService {
  client: OAuth2Client;

  constructor(private mailerService: MailService) {
    this.client = new OAuth2Client(
      process.env.GOOGLE_CLIENT_ID,
      process.env.GOOGLE_CLIENT_SECRET,
      process.env.GOOGLE_REDIRECT_URI,
    );
  }

  async loginWithShibboleth(
    mail: string,
    role: string,
    givenName: string,
    lastName: string,
    organizationId: number,
  ): Promise<number> {
    try {
      const user = await UserModel.findOne({ email: mail });

      if (user && user.password) {
        throw new BadRequestException(
          'User collisions with legacy account are not allowed',
        );
      }

      if (user && user.accountType !== AccountType.SHIBBOLETH) {
        throw new BadRequestException(
          'User collisions with other account types are not allowed',
        );
      }

      if (user) {
        return user.id;
      }

      if (!user) {
        const newUser = await UserModel.create({
          email: mail,
          firstName: givenName,
          lastName: lastName,
          accountType: AccountType.SHIBBOLETH,
          emailVerified: true,
        }).save();

        const userId = newUser.id;

        await OrganizationUserModel.create({
          organizationId,
          userId: userId,
          role: OrganizationRole.MEMBER,
        }).save();

        await ChatTokenModel.create({
          user: newUser,
          token: v4(),
        }).save();

        return userId;
      }

      throw new InternalServerErrorException('Unexpected error');
    } catch (err) {
      throw new BadRequestException(err.message);
    }
  }

  async loginWithGoogle(
    auth_code: string,
    organizationId: number,
  ): Promise<number> {
    try {
      const { tokens } = await this.client.getToken(auth_code);

      const ticket = await this.client.verifyIdToken({
        idToken: tokens.id_token,
        audience: `${process.env.GOOGLE_CLIENT_ID}.apps.googleusercontent.com`,
      });

      const payload = ticket.getPayload();

      if (!payload.email_verified) {
        throw new BadRequestException('Email not verified');
      }

      const user = await UserModel.findOne({ email: payload.email });

      if (user && user.password) {
        throw new BadRequestException(
          'User collisions with legacy account are not allowed',
        );
      }

      if (user && user.accountType !== AccountType.GOOGLE) {
        throw new BadRequestException(
          'User collisions with other account types are not allowed',
        );
      }

      if (user) {
        return user.id;
      }

      if (!user) {
        const newUser = await UserModel.create({
          email: payload.email,
          firstName: payload.given_name,
          lastName: payload.family_name,
          photoURL: payload.picture,
          accountType: AccountType.GOOGLE,
          emailVerified: true,
        }).save();

        const userId = newUser.id;

        await OrganizationUserModel.create({
          organizationId,
          userId: userId,
        }).save();

        await ChatTokenModel.create({
          user: newUser,
          token: v4(),
        }).save();

        return userId;
      }

      throw new InternalServerErrorException('Unexpected error');
    } catch (err) {
      throw new BadRequestException(err.message);
    }
  }

  async register(
    firstName: string,
    lastName: string,
    email: string,
    password: string,
    sid: number,
    organizationId: number,
  ): Promise<number> {
    try {
      const user = await UserModel.findOne({ email });

      if (user) {
        throw new BadRequestException('Email already exists');
      }

      let newUser: UserModel;

      const salt = await bcrypt.genSalt(10);
      const hashedPassword = await bcrypt.hash(password, salt);

      if (sid == -1) {
        newUser = await UserModel.create({
          courses: [],
          email,
          firstName,
          lastName,
          password: hashedPassword,
          hideInsights: [],
        }).save();
      } else {
        newUser = await UserModel.create({
          courses: [],
          email,
          firstName,
          lastName,
          password: hashedPassword,
          sid,
          hideInsights: [],
        }).save();
      }

      const createdAt: number = parseInt(new Date().getTime().toString());
      const token: string = this.generateToken(8);

      await UserTokenModel.create({
        user: newUser,
        token: token,
        created_at: createdAt,
        expires_at: createdAt + 1000 * 60 * 60 * 24,
      }).save();

      await this.mailerService.sendUserVerificationCode(token, email);

      const userId = newUser.id;

      await OrganizationUserModel.create({
        organizationId,
        userId,
        role: OrganizationRole.MEMBER,
      }).save();

      await ChatTokenModel.create({
        user: newUser,
        token: v4(),
      }).save();

      return userId;
    } catch (err) {
      throw new BadRequestException(err.message);
    }
  }

  async studentIdExists(sid: number, oid: number): Promise<boolean> {
    const user = await UserModel.findOne({
      where: { sid },
      relations: ['organizationUser'],
    });
    return user && user.organizationUser.organizationId === oid ? true : false;
  }

  async createPasswordResetToken(user: UserModel): Promise<string> {
    const token = this.generateToken(12).toLowerCase();
    const createdAt = parseInt(new Date().getTime().toString());

    await UserTokenModel.create({
      user,
      token,
      created_at: createdAt,
      expires_at: createdAt + 1000 * 60 * 60 * 24,
      token_type: TokenType.PASSWORD_RESET,
    }).save();

    return token;
  }

  private generateToken(length: number): string {
    const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
    let token = '';
    for (let i = 0; i < length; i++) {
      const randomIndex = Math.floor(Math.random() * characters.length);
      token += characters.charAt(randomIndex);
    }
    return token;
  }
}
