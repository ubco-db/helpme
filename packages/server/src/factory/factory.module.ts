import { Module } from '@nestjs/common';
import { FactoryService } from './factory.service';

@Module({
  providers: [FactoryService],
  exports: [FactoryService],
})
export class FactoryModule {}
