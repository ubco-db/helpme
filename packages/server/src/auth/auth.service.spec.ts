import { Test, TestingModule } from '@nestjs/testing';
import { AuthService } from './auth.service';
import { TestConfigModule, TestTypeOrmModule } from '../../test/util/testUtils';
import { UserModel } from 'profile/user.entity';
import { DataSource } from 'typeorm';
import {
  initFactoriesFromService,
  OrganizationFactory,
  OrganizationUserFactory,
} from '../../test/util/factories';
import { AccountType } from '@koh/common';
import { OrganizationUserModel } from 'organization/organization-user.entity';
import { MailService } from 'mail/mail.service';
import { MailModule } from 'mail/mail.module';
import { FactoryModule } from 'factory/factory.module';
import { FactoryService } from 'factory/factory.service';

// Extend the OAuth2Client mock with additional methods
jest.mock('google-auth-library', () => {
  const actualLibrary = jest.requireActual('google-auth-library');

  class MockOAuth2Client extends actualLibrary.OAuth2Client {
    async getToken(code: string): Promise<any> {
      if (code === 'valid_code') {
        return Promise.resolve({ tokens: { id_token: 'valid_token' } });
      } else {
        return Promise.resolve({ tokens: { id_token: 'mocked_token' } });
      }
    }

    async verifyIdToken(options: any): Promise<any> {
      if (options.idToken !== 'valid_token') {
        return Promise.resolve({
          getPayload: () => ({
            email_verified: false,
            email: 'mocked_email@example.com',
            given_name: 'John',
            family_name: 'Doe',
            picture: 'mocked_picture_url',
          }),
        });
      } else {
        return Promise.resolve({
          getPayload: () => ({
            email_verified: true,
            email: 'mocked_email@example.com',
            given_name: 'John',
            family_name: 'Doe',
            picture: 'mocked_picture_url',
          }),
        });
      }
    }
  }

  return {
    OAuth2Client: MockOAuth2Client,
  };
});

class MockMailService {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  async sendUserVerificationCode(
    code: string,
    receiver: string,
  ): Promise<void> {
    return;
  }
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  async sendPasswordResetEmail(receiver: string, url: string): Promise<void> {
    return;
  }
}

describe('AuthService', () => {
  let service: AuthService;
  let dataSource: DataSource;
  let mailService: MailService;

  beforeAll(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [TestTypeOrmModule, TestConfigModule, FactoryModule, MailModule],
      providers: [
        AuthService,
        { provide: MailService, useClass: MockMailService },
      ],
    }).compile();

    service = module.get<AuthService>(AuthService);
    mailService = module.get<MailService>(MailService);
    dataSource = module.get<DataSource>(DataSource);

    // Grab FactoriesService from Nest
    const factories = module.get<FactoryService>(FactoryService);
    // Initialize the named exports to point to the actual factories
    initFactoriesFromService(factories);
  });

  afterAll(async () => {
    await dataSource.destroy();
  });

  beforeEach(async () => {
    await dataSource.synchronize(true);
  });

  describe('loginWithShibboleth', () => {
    it('should throw an error when user already exists with password', async () => {
      await UserModel.create({
        email: 'mocked_email@example.com',
        password: 'test_password',
      }).save();

      await expect(
        service.loginWithShibboleth(
          'mocked_email@example.com',
          'John',
          'Doe',
          1,
        ),
      ).rejects.toThrowError(
        'A non-SSO account already exists with this email. Please login with your email and password instead.',
      );
    });

    it('should throw an error when user already exists with other account type', async () => {
      await UserModel.create({
        email: 'mocked_email@example.com',
        accountType: AccountType.GOOGLE,
      }).save();

      await expect(
        service.loginWithShibboleth(
          'mocked_email@example.com',
          'John',
          'Doe',
          1,
        ),
      ).rejects.toThrowError(
        'A non-SSO account already exists with this email. Please login with your email and password instead.',
      );
    });

    it('should return user id when user already exists without password', async () => {
      const user = await UserModel.create({
        email: 'mocked_email@example.com',
        accountType: AccountType.SHIBBOLETH,
      }).save();

      const userId = await service.loginWithShibboleth(
        'mocked_email@example.com',
        'John',
        'Doe',
        1,
      );
      expect(userId).toEqual(user.id);
    });

    it('should create a new user when user does not exist', async () => {
      const organization = await OrganizationFactory.create();

      const userId = await service.loginWithShibboleth(
        'mocked_email@example.com',
        'John',
        'Doe',
        organization.id,
      );
      const user = await UserModel.findOne({
        where: {
          id: userId,
        },
      });
      expect(user).toMatchSnapshot();
    });
  });

  describe('loginWithGoogle', () => {
    it('should throw an error when email is not verified', async () => {
      await expect(service.loginWithGoogle('invalid_token', 1)).rejects.toThrow(
        'Email not verified',
      );
    });

    it('should throw an error when user already exists with password', async () => {
      const organization = await OrganizationFactory.create();
      const user = await UserModel.create({
        email: 'mocked_email@example.com',
        password: 'test_password',
      }).save();
      await OrganizationUserFactory.create({
        organizationUser: user,
        organization: organization,
      });

      await expect(
        service.loginWithGoogle('valid_code', organization.id),
      ).rejects.toThrowError(
        'A non-SSO account already exists with this email. Please login with your email and password instead.',
      );
    });

    it('should throw an error when user already exists with other account type', async () => {
      const organization = await OrganizationFactory.create();
      const user = await UserModel.create({
        email: 'mocked_email@example.com',
        accountType: AccountType.SHIBBOLETH,
      }).save();
      await OrganizationUserFactory.create({
        organizationUser: user,
        organization: organization,
      });

      await expect(
        service.loginWithGoogle('valid_code', organization.id),
      ).rejects.toThrowError(
        'A non-google account already exists with this email on HelpMe. Please try logging in with your email and password instead (or another SSO provider)',
      );
    });

    it('should return user id when user already exists without password', async () => {
      const organization = await OrganizationFactory.create();
      const user = await UserModel.create({
        email: 'mocked_email@example.com',
        accountType: AccountType.GOOGLE,
      }).save();
      await OrganizationUserFactory.create({
        organizationUser: user,
        organization: organization,
      });

      const userId = await service.loginWithGoogle(
        'valid_code',
        organization.id,
      );
      expect(userId).toEqual(user.id);
    });

    it('should create a new user when user does not exist', async () => {
      const organization = await OrganizationFactory.create();

      const userId = await service.loginWithGoogle(
        'valid_code',
        organization.id,
      );
      const user = await UserModel.findOne({
        where: {
          id: userId,
        },
      });
      expect(user).toMatchSnapshot();
    });
  });

  describe('studentIdExists', () => {
    it('should return false when student id does not exist in organization', async () => {
      const organization = await OrganizationFactory.create();

      const result = await service.studentIdExists(-1, organization.id);
      expect(result).toBe(false);
    });

    it('should return false when studnet id exists but in different organization', async () => {
      const organization = await OrganizationFactory.create();
      const otherOrganization = await OrganizationFactory.create();

      const user = await UserModel.create({
        email: 'test@email.com',
        sid: 123456789,
      }).save();

      await OrganizationUserModel.create({
        userId: user.id,
        organizationId: otherOrganization.id,
      }).save();

      const result = await service.studentIdExists(user.sid, organization.id);
      expect(result).toBe(false);
    });

    it('should return true when student id exists in organization', async () => {
      const organization = await OrganizationFactory.create();

      const user = await UserModel.create({
        email: 'test@email.com',
        sid: 123456789,
      }).save();

      await OrganizationUserModel.create({
        userId: user.id,
        organizationId: organization.id,
      }).save();

      const result = await service.studentIdExists(user.sid, organization.id);
      expect(result).toBe(true);
    });
  });

  describe('register', () => {
    it('should throw an error when email already exists', async () => {
      await UserModel.create({
        email: 'existingEmail@mail.com',
      }).save();

      await expect(
        service.register(
          'John',
          'Doe',
          'existingEmail@mail.com',
          'password',
          -1,
          1,
        ),
      ).rejects.toThrowError('Email already exists');
    });

    it('should create a new user when email does not exist with empty sid', async () => {
      const organization = await OrganizationFactory.create();

      const userId = await service.register(
        'John',
        'Doe',
        'email@mail.com',
        'password',
        -1,
        organization.id,
      );

      const user = await UserModel.findOne({
        where: {
          id: userId,
        },
      });
      expect(userId == user.id).toBe(true);
    });

    it('should create a new user when email does not exist with sid', async () => {
      const organization = await OrganizationFactory.create();

      const userId = await service.register(
        'John',
        'Doe',
        'email@mail.com',
        'password',
        123456,
        organization.id,
      );

      const user = await UserModel.findOne({
        where: {
          id: userId,
        },
      });

      expect(userId == user.id).toBe(true);
      expect(user.sid).toBe(123456);
    });

    it('should throw an error when unexpected error occurs', async () => {
      await expect(
        service.register('John', 'Doe', 'email@mail.com', 'password', -1, 1),
      ).rejects.toThrowError;
    });
  });
});
