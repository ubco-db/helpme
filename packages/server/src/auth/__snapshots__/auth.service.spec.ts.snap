// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthService loginWithGoogle should create a new user when user does not exist 1`] = `
UserModel {
  "accountDeactivated": false,
  "accountType": "google",
  "defaultMessage": null,
  "desktopNotifsEnabled": false,
  "email": "mocked_email@example.com",
  "emailVerified": true,
  "firstName": "John",
  "hideInsights": null,
  "id": 1,
  "includeDefaultMessage": true,
  "lastName": "Doe",
  "name": "John Doe",
  "password": null,
  "photoURL": "mocked_picture_url",
  "readChangeLog": false,
  "sid": null,
  "userRole": "user",
}
`;

exports[`AuthService loginWithShibboleth should create a new user when user does not exist 1`] = `
UserModel {
  "accountDeactivated": false,
  "accountType": "shibboleth",
  "defaultMessage": null,
  "desktopNotifsEnabled": false,
  "email": "mocked_email@example.com",
  "emailVerified": true,
  "firstName": "John",
  "hideInsights": null,
  "id": 1,
  "includeDefaultMessage": true,
  "lastName": "Doe",
  "name": "John Doe",
  "password": null,
  "photoURL": null,
  "readChangeLog": false,
  "sid": null,
  "userRole": "user",
}
`;
